unit_tests/HttpClient.d

```diff
@@ -0,0 +1,326 @@
+// HTTP utils based on https://github.com/adamdruppe/arsd/blob/master/http.d
+// copyright Adam D. Ruppe
+module HttpClient;
+
+import std.socket;
+import std.string;
+import std.conv;
+static import std.algorithm;
+static import std.uri;
+
+ubyte[] getBinary(string url, string[string] cookies = null) {
+	auto hr = httpRequest("GET", url, null, cookies);
+	if(hr.code != 200)
+		throw new Exception(format("HTTP answered %d instead of 200 on %s", hr.code, url));
+	return hr.content;
+}
+
+struct HttpResponse {
+	int code;
+	string contentType;
+	string[string] cookies;
+	string[] headers;
+	ubyte[] content;
+}
+
+struct UriParts {
+	string original;
+	string method;
+	string host;
+	ushort port;
+	string path;
+
+	bool useHttps;
+
+	this(string uri) {
+		original = uri;
+
+		if(uri[0 .. 8] == "https://")
+			useHttps = true;
+		else
+		if(uri[0..7] != "http://")
+			throw new Exception("You must use an absolute, http or https URL.");
+
+		version(with_openssl) {} else
+		if(useHttps)
+			throw new Exception("openssl support not compiled in try -version=with_openssl");
+
+		int start = useHttps ? 8 : 7;
+
+		auto posSlash = uri[start..$].indexOf("/");
+		if(posSlash != -1)
+			posSlash += start;
+
+		if(posSlash == -1)
+			posSlash = uri.length;
+
+		auto posColon = uri[start..$].indexOf(":");
+		if(posColon != -1)
+			posColon += start;
+
+		if(useHttps)
+			port = 443;
+		else
+			port = 80;
+
+		if(posColon != -1 && posColon < posSlash) {
+			host = uri[start..posColon];
+			port = to!ushort(uri[posColon+1..posSlash]);
+		} else
+			host = uri[start..posSlash];
+
+		path = uri[posSlash..$];
+		if(path == "")
+			path = "/";
+	}
+}
+
+HttpResponse httpRequest(string method, string uri, const(ubyte)[] content = null, string[string] cookies = null, string[] headers = null) {
+	import std.socket;
+
+	auto u = UriParts(uri);
+	// auto f = openNetwork(u.host, u.port);
+	auto f = new TcpSocket();
+	f.connect(new InternetAddress(u.host, u.port));
+
+	void delegate(string) write = (string d) {
+		f.send(d);
+	};
+
+	char[4096] readBuffer; // rawRead actually blocks until it can fill up the whole buffer... which is broken as far as http goes so one char at a time i guess. slow lol
+	char[] delegate() read = () {
+		size_t num = f.receive(readBuffer);
+		return readBuffer[0..num];
+	};
+
+	version(with_openssl) {
```

Comment:
Hmm, so it doesn't support SSL out of the box? But I guess since the tests are passing we don't need that.